### PR TITLE
fix: expand ReentrantLock scope in MonitorThreadContainer for better …

### PR DESCRIPTION
**Pull request by aaronchung-bitquill**
…thread safety

### Summary
Remove race condition in `MonitorThreadContainer` between `getInstance()` and `releaseInstance()` methods.

### Description

In the `MonitorThreadContainer`, there are possible race conditions between the `getInstance()` and `releaseInstance()` methods. 

In one of the race conditions is in the `getInstance()`. 
```
...
      } finally {
        LOCK_OBJECT.unlock();
      }
    }
    CLASS_USAGE_COUNT.getAndIncrement();
    return singleton;
  }
```
Just after the lock object is unlocked, is an opportunity for a race condition with another thread in the `releaseInstance()` method. 
```
...
    if (CLASS_USAGE_COUNT.decrementAndGet() <= 0) {
      LOCK_OBJECT.lock();
      try {
        if (singleton != null) {
          singleton.releaseResources();
          singleton = null;
          CLASS_USAGE_COUNT.set(0);
...
```
If the locked block is entered, a race condition may result.  `getInstance()` may  return a null `singleton` which will be followed by a NPE. Or another scenario is that a non-null singleton will be returned, but `singleton.releaseResources()` is called so the `MonitorThreadContainer`'s `threadPool` will be shutdown and subsequent usage of the threadPool will result in a `RejectedExecutionException`.

### By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by aaronchung-bitquill** [comment]
Just realized that perhaps using `synchronized` on both methods may be a better solution.

File: wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
+/*
```
Comment (brunos-bq) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java:1]: NIT: can we rename it to ConcurrencyEFMTest ?

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### :magic_wand: Added
 - Host Availability Strategy to help keep host health status up to date ([PR #530](https://github.com/awslabs/aws-advanced-jdbc-wrapper/pull/530)).
 
+### :bug: Fixed
+- Race conditions between `MonitorThreadContainer#getInstance()` and `MonitorThreadContainer#releaseInstance()` ([PR #601](https://github.com/awslabs/aws-advanced-jdbc-wrapper/pull/601)).
```
Comment (brunos-bq) [inline CHANGELOG.md:11]: NIT: Fixed a case of race condition between...

File: wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    Runnable getInstanceTask = () -> {
+      MonitorThreadContainer.getInstance(mockExecutorServiceInitializer);
+    };
+
+    Runnable releaseInstanceTask = MonitorThreadContainer::releaseInstance;
+
+    MonitorThreadContainer.getInstance(mockExecutorServiceInitializer);
+
+    ExecutorService executorService = Executors.newFixedThreadPool(2);
+    executorService.execute(getInstanceTask);
```
Comment (brunos-bq) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java:63]: should we execute getInstance and releaseInstance more times here?

**Comment by davecramer** [comment]
Instead of using synchronized it would be better to use the re-entrant lock before the first check. synchronized is still a heavier lock.

File: wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java (hunk 1 above)
Comment (aaronchung-bitquill) [reply wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java:1]: `ConcurrencyEFMTest` may be too general. There is also another test class `MultiThreadedDefaultMonitorServiceTest` that also tests concurrency in EFM.

Comment (aaronchung-bitquill) [reply wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java:1]: Is the nit b/c `MultiThreadedMonitorThreadContainerTest` is extremely verbose? :D 

I was thinking of replacing `MultiThreaded` with `Concurrency` so the test class name would be `ConcurrencyMonitorTheadContainerTest` but that still seems a bit wordy. 

What do you think?

File: CHANGELOG.md (hunk 2 above)
Comment (aaronchung-bitquill) [reply CHANGELOG.md:11]: None of the Fixed changelog items start with "Fixed ...." and instead either state the bug, or what was done to fix the bug. I did this once and was called out and IIRC the reason was that "Fixed..."  is redundant with the header.

File: wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java (hunk 3 above)
Comment (aaronchung-bitquill) [reply wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java:69]: Actually, since this test a `@RepeatedTest`, I think this should suffice. Is that ok with you?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/efm/MonitorThreadContainer.java [hunk 4]
```diff
@@ -54,41 +54,35 @@ public static MonitorThreadContainer getInstance() {
...
-      LOCK_OBJECT.lock();
-      try {
-        if (singleton != null) {
-          singleton.releaseResources();
-          singleton = null;
-          CLASS_USAGE_COUNT.set(0);
-        }
-      } finally {
-        LOCK_OBJECT.unlock();
+    LOCK_OBJECT.lock();
```
Comment (brunos-bq) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/efm/MonitorThreadContainer.java:77]: just double checking - do we enter the lock section even if singleton is null when method is called?

Comment (aaronchung-bitquill) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/efm/MonitorThreadContainer.java:77]: Yes. I originally thought there could b a race condition in the current code where after the null check, and before it returns, another thread could call getInstance, which would make the singleton not null, and the original thread thats calling releaseInstance wouldnt know about that.
But now im wondering if thats an issue at all.

File: CHANGELOG.md (hunk 2 above)
Comment (aaronchung-bitquill) [reply CHANGELOG.md]: Done!

File: wrapper/src/main/java/software/amazon/jdbc/plugin/efm/MonitorThreadContainer.java (hunk 4 above)
Comment (aaronchung-bitquill) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/efm/MonitorThreadContainer.java:77]: I will add the null check and early return back. 
This isnt the area that we've seen issues with, so I think the safer approach here is to leave it as it was.

**Review by crystall-bitquill — APPROVED** [review]

File: wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.RepeatedTest;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+public class MultiThreadedMonitorThreadContainerTest {
+
+  @Mock
+  ExecutorServiceInitializer mockExecutorServiceInitializer;
```
Comment (karenc-bq) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java:34]:
```suggestion
  @Mock ExecutorServiceInitializer mockExecutorServiceInitializer;
```

File: wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  @Mock
+  ExecutorService mockExecutorService;
```
Comment (karenc-bq) [inline wrapper/src/test/java/software/amazon/jdbc/plugin/efm/MultiThreadedMonitorThreadContainerTest.java:36]:
```suggestion
  @Mock ExecutorService mockExecutorService;
```

**Review by karenc-bq — APPROVED** [review]
